packages/@aws-cdk/aws-servicecatalog/lib/product.ts

```diff
@@ -202,20 +208,45 @@ export class CloudFormationProduct extends Product {
     }
   }
 
-  private renderProvisioningArtifacts(
-    props: CloudFormationProductProps): CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] {
-    return props.productVersions.map(productVersion => {
+  private renderProvisioningArtifacts(props: CloudFormationProductProps): CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] {
+    let productVersions: CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty[] = [];
+    for (const productVersion of props.productVersions) {
       const template = productVersion.cloudFormationTemplate.bind(this);
-      InputValidator.validateUrl(this.node.path, 'provisioning template url', template.httpUrl);
-      return {
+      let httpUrl = template.httpUrl;
+      switch (template.templateType) {
+        case TemplateType.PRODUCT_STACK:
+          if (template.productVersionDetails) {
+            template.productVersionDetails.productPathUniqueId = this.productPathUniqueId;
+            template.productVersionDetails.productVersionName = productVersion.productVersionName;
+            template.productVersionDetails.retentionStrategy = template.retentionStrategy;
+          }
+          break;
+        case TemplateType.PRODUCT_STACK_CONTEXT:
+          const templateFileKey = `${this.productPathUniqueId}.${template.productVersionDetails?.productStackId}.${productVersion.productVersionName}.product.template.json`;
+          const templateFilePath = path.join(PRODUCT_STACK_SNAPSHOT_DIRECTORY, templateFileKey);
+          if (!fs.existsSync(templateFilePath)) {
+            throw new Error(`Template ${templateFileKey} cannot be found in ${PRODUCT_STACK_SNAPSHOT_DIRECTORY}`);
+          }
+          httpUrl = new s3_assets.Asset(this, `Template${hashValues(templateFileKey)}`, {
+            path: templateFilePath,
+          }).httpUrl;
+          break;
+        default:
+          break;
+      }
+
+      InputValidator.validateUrl(this.node.path, 'provisioning template url', httpUrl);
+      productVersions.push({
         name: productVersion.productVersionName,
         description: productVersion.description,
         disableTemplateValidation: productVersion.validateTemplate === false ? true : false,
         info: {
-          LoadTemplateFromURL: template.httpUrl,
+          LoadTemplateFromURL: httpUrl,
         },
-      };
-    });
+      },
+      );
```

Comment:
my suggested change should have been:
```suggestion
      });
```